File: 1411. number-of-ways-to-paint-n-3-grid.py
class Solution:
    def numOfWays(self, n: int) -> int:
        MOD = 10**9 + 7
        A = B = 6

        for _ in range(2, n + 1):
            A, B = (2*A + 2*B) % MOD, (2*A + 3*B) % MOD

        return (A + B) % MOD

# Example usage
sol = Solution()
print(sol.numOfWays(1))  # Output: 12	
print(sol.numOfWays(5000))  # Output: 30228214

File: 1411. number-of-ways-to-paint-n-3-grid.java
class Solution {
    public int numOfWays(int n) {
        int MOD = 1000000007;
        long A = 6;
        long B = 6;

        for (int i = 2; i <= n; i++) {
            long newA = (2 * A + 2 * B) % MOD;
            long newB = (2 * A + 3 * B) % MOD;
            A = newA;
            B = newB;
        }

        return (int)((A + B) % MOD);
    }
    public static void main(String[] args) {
    Solution solution = new Solution();
    
    // Test cases
    System.out.println("n = 1: " + solution.numOfWays(1)); // Expected: 12
    System.out.println("n = 2: " + solution.numOfWays(2)); // Expected: 54
    System.out.println("n = 3: " + solution.numOfWays(5000)); // Expected: 30228214
}
}